### PR TITLE
Add interpreter for NegOp

### LINKED ISSUE #178: Add interpreter for NegOp

### PR DESCRIPTION
**Pull request by ghpvnist**
closes #178

### CODE REVIEW

File: stablehlo/reference/Element.cpp [hunk 1]
```diff
@@ -200,6 +200,12 @@ Element Element::operator+(const Element &other) const {
       });
 }
 
+Element Element::operator-() const {
+  return map(
+      *this, [&](APInt api) { return -api; }, [&](APFloat apf) { return -apf; },
```
Comment (burmako) [inline stablehlo/reference/Element.cpp:205]: Let's rename `api`, `apf` and `complex` to `val` similarly to how similar variables are called in `ceil` and `floor`?

File: stablehlo/tests/interpret_negate.mlir [hunk 2]
```diff
@@ -0,0 +1,247 @@
+// RUN: stablehlo-interpreter --interpret -split-input-file %s | FileCheck %s
+
+// CHECK-LABEL: Evaluated results of function: negate_op_test_si4
+func.func @negate_op_test_si4() -> tensor<5xi4> {
+  %0 = stablehlo.constant dense<[-8, -1, 0, 1, 7]> : tensor<5xi4>
+  %1 = stablehlo.negate %0 : tensor<5xi4>
+  func.return %1 : tensor<5xi4>
+  // CHECK-NEXT: tensor<5xi4>
+  // CHECK-NEXT: -8 : i4
```
Comment (burmako) [inline stablehlo/tests/interpret_negate.mlir:9]: Sounds like the spec might need an update? Currently, it says that negation of `-8` should be either `7` or `-9`.

**Review by burmako — CHANGES_REQUESTED** [review]

File: stablehlo/reference/Interpreter.cpp [hunk 3]
```diff
@@ -97,6 +97,10 @@ llvm::Expected<SmallVector<Tensor>> eval(func::FuncOp func,
       Tensor runtimeOperand = fetchOperand(floorOp.operand());
       Tensor runtimeResult = eval(floorOp, runtimeOperand);
       populateResults({runtimeResult});
+    } else if (auto negOp = dyn_cast<NegOp>(op)) {
+      Tensor runtimeOperand = fetchOperand(negOp.operand());
```
Comment (GleasonK) [inline stablehlo/reference/Interpreter.cpp:101]: Rebase this CL on main before submit, `operand()` will need to be updated to `getOperand()` with #203 submitted.

File: stablehlo/tests/interpret_negate.mlir (hunk 2 above)
Comment (sdasgup3) [reply stablehlo/tests/interpret_negate.mlir:9]: acknowledged.

File: stablehlo/reference/Interpreter.cpp (hunk 3 above)
Comment (ghpvnist) [reply stablehlo/reference/Interpreter.cpp:101]: Thanks for the heads up!

**Review by burmako — APPROVED** [review]

**Review by GleasonK — APPROVED** [review]

File: stablehlo/reference/Ops.cpp [hunk 4]
```diff
@@ -56,6 +56,14 @@ Tensor eval(FloorOp op, const Tensor &operand) {
   return result;
 }
 
+Tensor eval(NegOp op, const Tensor &operand) {
+  Tensor result(op.getType());
+  for (auto it = operand.index_begin(); it != operand.index_end(); ++it) {
+    result.set(*it, -operand.get(*it));
+  }
+  return result;
+}
```
Comment (joker-eph) [inline stablehlo/reference/Ops.cpp:65]: Drive by: seems like we're creating a new tensor unconditionally, but @sdasgup3 was telling me yesterday that Tensor is mutable. 
Seems like we could check if we're the only (or better: last) uses of the operand and mutate it in-place.

**Pull request merged**